### ISSUE TITLE
GpuIndexFlatL2 on multiple GPU

### ISSUE THREAD

**Issue by Ganstrich** [facebookresearch/faiss#222]
Hi,

I am trying to build a FlatL2 index on multiple gpus based from the examples in bench_gpu_1bn. The IVFFlat et IVPQ indexes work fine but I can't seem to find how to add iterations of my dataset to the index. The documentation states that the add_core and add_core_o functions work similarly to add_with_ids. However, add_with_ids on the FlatL2 doesn't seem to work. I'm giving to the function the same parameters as for IVFFlat except the precomputed coarse quantizer part since I'm working with FlatL2. Am I missing something ?

Thanks in advance !

**Comment by jegou** [comment]
FlatL2 does not support natively add_with_ids because it does not store ids by default. You can add this feature to this index by using the encapsulating IndexIDMap:
https://github.com/facebookresearch/faiss/wiki/Pre--and-post-processing

**Comment by mdouze** [comment]
Hi 
The bench_gpu_1bn is way too complex to make a FlatL2 index. You can just do

```
index = index_cpu_to_all_gpu(faiss.index_factory("IDMap,FlatL2"))


```

where 

```
def index_cpu_to_all_gpus(index, co=None):
    ngpu = faiss.get_num_gpus()
    res = [faiss.StandardGpuResources() for i in range(ngpu)]
    index2 = faiss.index_cpu_to_gpu_multiple_py(res, index, co)
    index2.dont_dealloc = res
    return index2
```

**Comment by Ganstrich** [comment]
Hi, 
Thanks for your answers. I have tried @mdouze solution and got an error on wrong number of arguments. Looked at usage of index_factory and noticed it expects an int as first arg, so I passed the dimension of vectors. But now I got an "Error in virtual faiss : clone not supported for this type of Index".

**Comment by wickedfoo** [comment]
It's on my todo list to add native multi-GPU support to the indexes, starting with GpuIndexFlat, so you don't need to wrap them in a different index.

**Comment by mdouze** [comment]
This has been fixed in 1.4.0.

**Comment by jonbesga** [comment]
Is it possible to do this directly from `faiss.GpuIndexFlatL2` without having to create a CPU index first and transfer it to the GPU?
